### PR TITLE
Development

### PR DESCRIPTION
**Pull request by nboj**

### CODE REVIEW

File: src/deimoslang/vm.py [hunk 1]
```diff
@@ -551,6 +571,20 @@ async def step(self):
             case InstructionKind.label | InstructionKind.nop:
                 self._ip += 1
 
+            case InstructionKind.push_stack:
+                self._callstack.append(None)
```
Comment (SirOlaf) [inline src/deimoslang/vm.py:575]: Rename _callstack to _stack if you're gonna use it for this too

File: src/deimoslang/vm.py [hunk 2]
```diff
@@ -270,6 +286,10 @@ async def _eval_expression(self, kind:EvalKind, client: Client):
                 return await client.stats.current_gold()
             case EvalKind.max_gold:
                 return await client.stats.base_gold_pouch()
+            case EvalKind.windowtext:
+                path = eval.args[0]
+                assert(type(path)==list)
+                return await (await get_window_from_path(client.root_window, path)).read_wide_string_from_offset(616)
```
Comment (SirOlaf) [inline src/deimoslang/vm.py:292]: Why a numeric offset instead of a read field?

**Review by SirOlaf — CHANGES_REQUESTED** [review]

File: src/deimoslang/types.py [hunk 3]
```diff
@@ -192,6 +193,10 @@ class EquivalentExpression(BinaryExpression):
     def __repr__(self) -> str:
         return f"EquivalentE({self.lhs}, {self.rhs})"
 
+class ContainsExpression(BinaryExpression):
```
[the remark this answers is not in the document]
Comment (nboj) [reply src/deimoslang/types.py]: changed to: ContainsStringExpression

File: src/deimoslang/ir.py [hunk 4]
```diff
@@ -187,15 +191,21 @@ def compile_call(self, call: CallStmt):
...
                 self.prep_expression(expr.rhs)
             case ReadVarExpr():
                 if isinstance(expr.loc, SymExpression):
                     expr.loc = StackLocExpression(self.stack_loc(expr.loc.sym))
                 else:
                     raise CompilerError(f"Malformed ReadVarExpr: {expr}")
-            case NumberExpression():
+            case SelectorGroup():
+                self.prep_expression(expr.expr)
+            case ReadVarExpr():
```
[the remark this answers is not in the document]
Comment (nboj) [reply src/deimoslang/ir.py]: removed.

File: src/deimoslang/vm.py (hunk 1 above)
Comment (nboj) [reply src/deimoslang/vm.py]: done

File: src/deimoslang/vm.py (hunk 2 above)
Comment (nboj) [reply src/deimoslang/vm.py]: added try/catch and maybe_text

File: src/deimoslang/sem.py [hunk 5]
```diff
@@ -39,6 +39,24 @@ def __init__(self, stmts: list[Stmt]):
         self._block_defs: list[BlockDefStmt] = []
         self._stmts = stmts
 
+        self._block_nesting_level = 0
+        self._loop_nesting_level = 0
+
+    def open_block(self):
+        self.open_scope()
+        self._block_nesting_level+=1
```
[the remark this answers is not in the document]
Comment (nboj) [reply src/deimoslang/sem.py:50]: made a stack

File: src/deimoslang/vm.py [hunk 6]
```diff
@@ -241,18 +240,35 @@ async def eval(self, expression: Expression, client: Client | None = None):
...
+                return lhs - rhs
+            case ContainsStringExpression():
+                lhs = await self.eval(expression.lhs, client)
+                rhs = await self.eval(expression.rhs, client)
+                return (rhs in lhs) #type: ignore
             case _:
                 raise VMError(f"Unimplemented expression type: {expression}")
 
-    async def _eval_expression(self, kind:EvalKind, client: Client):
+    async def _eval_expression(self, eval:Eval, client: Client):
```
Comment (SirOlaf) [inline src/deimoslang/vm.py:270]:
```suggestion
    async def _eval_expression(self, eval: Eval, client: Client):
```

File: src/deimoslang/vm.py [hunk 7]
```diff
@@ -551,6 +581,20 @@ async def step(self):
             case InstructionKind.label | InstructionKind.nop:
                 self._ip += 1
 
+            case InstructionKind.push_stack:
+                self._stack.append(None)
+                self._ip += 1
+
+            case InstructionKind.write_stack:
+                assert(instruction.data!=None)
```
Comment (SirOlaf) [inline src/deimoslang/vm.py:589]:
```suggestion
                assert(instruction.data != None)
```

File: src/deimoslang/vm.py [hunk 8]
```diff
@@ -480,18 +510,18 @@ async def step(self):
                     self._ip += instruction.data[1]
 
             case InstructionKind.call:
-                self._callstack.append(self._ip + 1)
-                jump = instruction.data  
-                self._ip += jump # type: ignore
+                assert(type(instruction.data)==int)
```
Comment (SirOlaf) [inline src/deimoslang/vm.py:513]:
```suggestion
                assert(type(instruction.data) == int)
```

File: src/deimoslang/vm.py [hunk 9]
```diff
@@ -270,6 +286,20 @@ async def _eval_expression(self, kind:EvalKind, client: Client):
                 return await client.stats.current_gold()
             case EvalKind.max_gold:
                 return await client.stats.base_gold_pouch()
+            case EvalKind.windowtext:
+                path = eval.args[0]
+                assert(type(path)==list)
```
Comment (SirOlaf) [inline src/deimoslang/vm.py:291]:
```suggestion
                assert(type(path) == list)
```

File: src/deimoslang/vm.py [hunk 10]
```diff
@@ -241,18 +240,35 @@ async def eval(self, expression: Expression, client: Client | None = None):
...
             case SelectorGroup():
                 players = self._select_players(expression.players)
                 expr = expression.expr
                 for player in players:
                     if not await self.eval(expr, player):
                         return False
                 return True
+            case ReadVarExpr():
+                loc = await self.eval(expression.loc)
+                assert(loc!=None and type(loc)==int)
```
Comment (SirOlaf) [inline src/deimoslang/vm.py:253]:
```suggestion
                assert(loc != None and type(loc) == int)
```

File: src/deimoslang/sem.py [hunk 11]
```diff
@@ -39,6 +39,28 @@ def __init__(self, stmts: list[Stmt]):
...
+    def open_block(self):
+        self.open_scope()
+        self._loop_nesting_stack.append(self._loop_nesting_level)
+        self._loop_nesting_level = 0
+        self._block_nesting_level+=1
+
+    def close_block(self):
+        self.close_scope()
+        self._loop_nesting_level = self._loop_nesting_stack.pop()
+        self._block_nesting_level-=1
```
Comment (SirOlaf) [inline src/deimoslang/sem.py:55]:
```suggestion
        self._block_nesting_level -= 1
```

File: src/deimoslang/sem.py [hunk 12]
```diff
@@ -39,6 +39,28 @@ def __init__(self, stmts: list[Stmt]):
...
+
+    def open_loop(self):
+        self.open_scope()
+        self._loop_nesting_level+=1
```
Comment (SirOlaf) [inline src/deimoslang/sem.py:59]:
```suggestion
        self._loop_nesting_level += 1
```

File: src/deimoslang/sem.py [hunk 13]
```diff
@@ -39,6 +39,28 @@ def __init__(self, stmts: list[Stmt]):
...
+
+    def close_loop(self):
+        self.close_scope()
+        self._loop_nesting_level-=1
```
Comment (SirOlaf) [inline src/deimoslang/sem.py:63]:
```suggestion
        self._loop_nesting_level -= 1

```

File: src/deimoslang/sem.py [hunk 14]
```diff
@@ -39,6 +39,28 @@ def __init__(self, stmts: list[Stmt]):
...
 
+        self._block_nesting_level = 0
+        self._loop_nesting_level = 0
+        self._loop_nesting_stack = []
+
+    def open_block(self):
+        self.open_scope()
+        self._loop_nesting_stack.append(self._loop_nesting_level)
+        self._loop_nesting_level = 0
+        self._block_nesting_level+=1
```
Comment (SirOlaf) [inline src/deimoslang/sem.py:50]:
```suggestion
        self._block_nesting_level += 1
```

File: src/deimoslang/parser.py [hunk 15]
```diff
@@ -679,6 +688,14 @@ def parse_stmt(self) -> Stmt:
                             elif_body_stack[-1].branch_false = StmtList([elif_stmt])
                         elif_body_stack.append(elif_stmt)
                 return IfStmt(expr, true_body, else_body)
+            case TokenKind.keyword_break:
+                self.i+=1
+                self.end_line()
+                return BreakStmt();
+            case TokenKind.keyword_return:
+                self.i+=1
```
Comment (SirOlaf) [inline src/deimoslang/parser.py:696]:
```suggestion
                self.i += 1
```

File: src/deimoslang/parser.py [hunk 16]
```diff
@@ -679,6 +688,14 @@ def parse_stmt(self) -> Stmt:
                             elif_body_stack[-1].branch_false = StmtList([elif_stmt])
                         elif_body_stack.append(elif_stmt)
                 return IfStmt(expr, true_body, else_body)
+            case TokenKind.keyword_break:
+                self.i+=1
```
Comment (SirOlaf) [inline src/deimoslang/parser.py:692]:
```suggestion
                self.i += 1
```

File: src/deimoslang/parser.py [hunk 17]
```diff
@@ -279,6 +279,16 @@ def parse_command_expression(self) -> Expression:
 
                 # target == evaluated
                 return self.gen_equivalent_expression(target, evaluated, player_selector)
+            case TokenKind.command_expr_window_text:
+                self.i += 1
+                window_path = self.parse_window_path()
+                contains = self.consume_optional(TokenKind.contains)
+                target = self.expect_consume(TokenKind.string)
+                assert(type(window_path)==list and type(target.value)==str)
```
Comment (SirOlaf) [inline src/deimoslang/parser.py:287]:
```suggestion
                assert(type(window_path) == list and type(target.value) == str)
```

File: src/deimoslang/ir.py [hunk 18]
```diff
@@ -158,17 +159,20 @@ def process_labels(self, program: list[Instruction]):
                     offset = offsets[sym]
                     program[idx] = Instruction(instr.kind, offset-idx)
                 case InstructionKind.jump_if | InstructionKind.jump_ifn | InstructionKind.enter_until:
+                    assert(type(instr.data)==list)
```
Comment (SirOlaf) [inline src/deimoslang/ir.py:162]:
```suggestion
                    assert(type(instr.data) == list)
```

**Review by SirOlaf — CHANGES_REQUESTED** [review]

**Pull request merged**